cli: add support for `CLIPlugin` interface to extend command line through plugins

wihobbs opened the pull request:
Outstanding to-dos:
- [x] eliminate need for `parser` argument in `CLIPlugin.add_options` method per James's suggestion
- [x] move prefix `--ex-` from the `CLIPluginRegistry` to each individual plugin per Ryan's suggestion

wihobbs commented:
> move prefix --ex- from the CLIPluginRegistry to each individual plugin per Ryan's suggestion

To pass arbitrary `**kwargs` to argparse without documenting and enumerating all of them, I created a `CLIPluginOption` class, which accepts arbitrary `**kwargs` (eventually invalid ones will error out, when the parser is created). It was logical to contain the `prefix="--ex-"` option there, if one plugin wants to use different prefixes, and it was workable in Python. 

Marking this as ready for review.

grondo commented on src/bindings/python/flux/cli/plugin.py:
```diff
@@ -0,0 +1,180 @@
+            if "dest" not in kwargs:
+                kwargs["dest"] = name[2:].replace("-", "_")
+            self.name = name.replace("--", prefix)
+        self.__dict__.update(kwargs)
```
Just curious what you get using `self.__dict__` vs just the more explicit use of an attribute like `self.kwargs`?

wihobbs replied:
More confusion, as is always the goal. (Fixed)

grondo commented on src/bindings/python/flux/cli/plugin.py:
```diff
@@ -0,0 +1,180 @@
+    def add_option(self, option):
+        """Allow plugin to add additional options to itself
+
+        Plugins use:
+        >>> self.add_option(CLIPluginOption("--longopt", action="store_true", help="Help."))
```
I would slightly prefer if the user did not have to instantiate the `CLIPluginOption()`. You may have a good reason for this, but why not:

```Python
    def add_option(name, prefix="--ex-", **kwargs):
        self.options.append(CLIPluginOption(name, prefix, **kwargs))
```

This is  a bit more user friendly interface IMO and doesn't require the user to import or use the `CLIPluginOption` class in the common case. However, you may have tried this and run into some issue I don't see.
```

grondo commented on src/bindings/python/flux/cli/plugin.py:
```diff
@@ -0,0 +1,180 @@
+        else:
+            self.options.append(option)
+
+    def get_options(self):
+        """Return a dict of options to add to the command line"""
+        return self.options
```
This method could be dropped in preference for just using `plugin.options`?

grondo commented on src/bindings/python/flux/cli/base.py:
```diff
@@ -757,7 +758,11 @@ def __init__(self, prog, usage=None, description=None, exclude_io=False):
         self.parser = self.create_parser(prog, usage, description, exclude_io)
+        group = self.parser.add_argument_group("Options provided by plugins")
+        for option in self.plugins.get_options():
+            group.add_argument(option.__dict__.pop("name"), **option.__dict__)
```
Another spot where having explicit attributes in the CLIPluginOption class would be good. Then you could split the option name and kwargs into `option.name` and `option.kwargs` and this code would be cleaner as:
```python
        for option in self.plugins.get_options():
            group.add_argument(option.name, **option.kwargs)

grondo commented on src/bindings/python/flux/cli/base.py:
```diff
@@ -1086,6 +1094,8 @@ def jobspec_create(self, args):
             for arg in args.add_file:
                 self.handle_add_file_arg(jobspec, arg)
 
+        self.plugins.modify_jobspec(args, jobspec)
```
Suggestion: Add a comment here just to indicate what the call is allowing.

grondo commented on src/bindings/python/flux/cli/plugin.py:
```diff
@@ -57,7 +57,7 @@ class CLIPlugin(ABC):  # pragma no cover
     """
 
-    def __init__(self, prog):
+    def __init__(self, prog, version=None, prefix="--ex-"):
```
This change may be in the wrong commit?

grondo commented on t/t2717-python-cli-plugins.t:
```diff
@@ -0,0 +1,46 @@
+
+test_under_flux 4 job
+
+test_expect_success 'flux-alloc: base --help message is formatted correctly' '
+  flux alloc --help 
+'
```
This should check something ;-) Maybe that the plugin option section appears in the output.

grondo reviewed (commented):
Looking good! Thanks @wihobbs!

I made a first pass through with some fairly minor comments.

Regarding testing, let's make sure we're testing:

 - failure modes, e.g. plugins that add a conflicting option. Conflicting options with different prefixes (which I think would result in the same "dest")
 - use of a different prefix works
 - use of an empty or None prefix works

wihobbs commented on src/bindings/python/flux/cli/base.py:
```diff
@@ -757,7 +758,11 @@ def __init__(self, prog, usage=None, description=None, exclude_io=False):
         self.exitcode = 0
         self.progress = None
         self.watcher = None
+        self.plugins = CLIPluginRegistry().load_plugins(prog)
```
@grondo What would you think of instead

```
self.plugins = CLIPluginRegistry(prog)
```
and the registry calling `load_plugins` as part of init? (or doing the same things but eliminating that function)

wihobbs replied:
We'd have to use `__new__` instead of `__init__` to return an instance of the class, perhaps that's why you went with the `load_plugins()` function instead in the first prototype.

wihobbs replied:
Ohh, I see why now. This is useful as-is in the validator. Ignore my noise!

grondo replied:
You make a good point, though. Since it is internal use only, we could eventually update the class initializer to work in the way you propose if someone thinks of a clean solution.

wihobbs replied:
I think I might have figured out how to do this, but I need to test it first.

grondo replied:
It does seem like you could just call `self.load_plugins(prog)` in the `__init__` function, once `prog` is added as an arg.

I guess it is just a style question at this point. The one time you may not want to do this is when you might add other setup or configuration methods to the class in the future. Then you want to be able to instantiate an object of the class _before_ loading plugins. This kind of reason is why I tend to first create classes that only set initial attribute values in a class initializer, and other methods to fully set up or configure the class.

wihobbs commented on src/bindings/python/flux/cli/plugin.py:
```diff
@@ -0,0 +1,180 @@
+
+    Other keyword arguments, except ``dest=`` are passed along to
+    ``ArgumentParser.add_argument`` unchanged. ``dest=`` is a special case,
+    where dashes are replaced with underscores.
```
> Do we fail in a predictable way if two options are registered from different plugins with the same name but different prefix?

I decided this should be a failure if the `kwargs["dest"]` option conflicts. Users have always had the option to provide their own dest, and perhaps there are some valid cases for having a plugin with the same `name` but different `dest`s. I'll push up how I did this soon.

cmoussa1 commented on src/bindings/python/flux/cli/plugin.py:
```diff
@@ -0,0 +1,180 @@
+##############################################################
+# Copyright 2024 Lawrence Livermore National Security, LLC
```
does this copyright year need to be updated to 2025?

grondo replied:
What we actually do is put the year the file was originally included in the source, then never worry about updating it. I don't have a problem with removing the year, but would like to get an ACK from @garlick since we haven't yet dropped the year everywhere else.

wihobbs replied:
Ok, that seems like maybe an out of scope conversation for this PR. I'll put 2025 in the header to be consistent and we can talk about dropping the year later.

cmoussa1 commented on t/cli-plugins/validateonly.py:
```diff
@@ -0,0 +1,42 @@
+        try:
+            verbosity = jobspec.attributes["system"]["shell"]["options"]["verbose"]
+            if not verbosity.isdigit():
+                raise ValueError(f"{self.name}: expected integer, got {verbosity}")
```
IMHO it could be useful to return the _type_ of `verbosity` instead of `verbosity` itself in this exception message, but you could probably take or leave this.

wihobbs replied:
Hmm, this plugin isn't being tested _at all_, that seems bad.

cmoussa1 commented on t/t2717-python-cli-plugins.t:
```diff
@@ -0,0 +1,62 @@
+  rm ../cli-plugins/doubleplugin.py
+'
+
+test_done
```
It looks like the indentation might be a little inconsistent (at least looking at it on GitHub). Is it possible some indentations are using spaces while others are using tabs?

cmoussa1 reviewed (commented):
This is looking pretty good! I have a couple of minor comments after a pass through.

wihobbs commented:
Thanks @cmoussa1! I believe I addressed all of your comments. I also fixed a failing test in `make distcheck` and added an additional test to make sure we can support plugins with the same name and different prefixes _if_ they specify a different `dest`.

grondo reviewed (commented):
Just did a quick run through and this is very close to ready to merge!

I'll actually check out the branch and do some hand testing then come back and do a another review.

cmoussa1 commented on src/bindings/python/flux/cli/plugin.py:
```diff
@@ -0,0 +1,180 @@
+        """Load all cli plugins from the standard path"""
+        for path in glob.glob(f"{self.plugindir}/*.py"):
+            self._add_plugins(import_path(path), program)
+        option_dests = []
```
You could use the [`.get()`](https://www.geeksforgeeks.org/python/python-dictionary-get-method/) method with a dict access that will return `None` (or I _believe_ a custom default) instead of raising a `KeyError`.

wihobbs replied:
Wow that is useful. 🤦

grondo replied:
I mean to continue using `if dest in option_dests`, it is the same code but (I presume) faster because `in` in a list would have to search the whole list, whereas `in` for a dict is just using the lookup.

wihobbs commented on t/t2717-python-cli-plugins.t:
```diff
@@ -0,0 +1,76 @@
+  grep "shell.options.verbose: expected integer, got <class" out5.err
+'
+test_expect_success 'flux-alloc: new plugin with the same name/dest as another fails' '
+  cat <<-EOF > ${SHARNESS_TEST_SRCDIR}/cli-plugins/doubleplugin.py && 
+from flux.cli.plugin import CLIPlugin 
+class BadPlugin(CLIPlugin):
+    def __init__(self, prog):
+        super().__init__(prog)
+        self.add_option("--gpumode", help="I want to mess up the gpumode")
+EOF
```
I actually have to drop these `cat` calls out of the tests because they're failing under `distcheck`. Will push up a proposed fix soon. I created new directories with failing plugins, but symlinked the working ones into that directory.

This is an unfortunate limitation of only having one statically defined directory for plugins, perhaps a future improvement could take a colon-delimited `PATH` or load plugins dynamically.

grondo replied:
Oh, I had forgotten the path can already be overidden using `FLUX_CLI_PLUGINPATH`. You could have tested these dynamically created plugins by setting that path. However, what you have is also fine and I do think we need to update this to take a colon delimited path, or support `XDG_CONFIG_HOME` and `XDG_CONFIG_DIRS` (need to ponder which is better). So keeping things as is for now is good.

wihobbs commented:
Thanks @grondo! Have fun on vacation! 🌴
